Selector access bug in legacy code generation. (#218)
ethereum/solidity-blog: _posts/2023-07-11-missing-side-effects-on-selector-access-bug.md

---
layout: post
published: true

> title: 'Bug when accessing the .selector member on complex expressions in legacy code generation'
> date: '2023-07-11'

Reviewer: IMO the title would sound better like this:

```suggestion
title: 'Bug in legacy code generation when accessing the .selector member on expressions with side-effects'
date: '2023-07-19'
```

Also, the likely date for the release is the next Wednesday. You can change that in the file name as well.

author: Solidity Team
category: Security Alerts
---

> On XXXX, a bug in the Solidity legacy code generator was found after further

Reviewer 2: Looking at our chats it looks like it was June 26

> On June 26, 2023, a bug in the Solidity legacy code generator was found after further

Reviewer: "Solidity legacy code generator" sounds a bit awkward. How about this?

```suggestion
On June 26, 2023, a bug in the legacy code generation pipeline of the Solidity compiler was found during
```

investigation of a security report related to the use of ``abi.decode`` with a ternary
expression as the type argument.

> The legacy code generation was not evaluating expressions when accessing the ``.selector``

Reviewer: ```suggestion
The legacy code generator was not evaluating expressions when accessing the ``.selector``
```

> The legacy code generation was not evaluating expressions when accessing the ``.selector``
> member on some complex expressions, like assignments and conditionals, while the ``via-ir`` code
> generation behave as expected and evaluates such expressions.
> This leads to a wrong compiler behavior in the legacy code generation.

Reviewer: ```suggestion
The legacy code generator was not evaluating complex expressions, like assignments, function calls or conditionals, whose  ``.selector`` was being accessed.
This led to the side-effects of such expressions not being executed, and therefore potentially incorrect behavior of contracts compiled using the legacy pipeline.
The via IR code generator behaves as expected and evaluates such expressions.
```

The bug was already present since the compiler version ``0.6.2``.
Despite this, as far as we know, it was never exploited or reported externally, and

> therefore we assigned it a severity of "low".

Reviewer: This justification seems a bit weak. I mean, the fact that it did not happen until now does not necessarily make it unlikely. There's more to it than that.

It's more that any expressions that do this can be seen as antipatterns and would likely be replaced with something much simpler. E.g. instead of using `f().g.selector` you're more likely to do it directly on the contract: `C.g.selector`. And potentially more sensible expressions like `f().selector` are not affected.

Author: > This justification seems a bit weak. I mean, the fact that it did not happen until now does not necessarily make it unlikely. There's more to it than that.

For sure. I will make the text clear about the anti-pattern.

> And potentially more sensible expressions like `f().selector` are not affected.

Not sure if I understood this sentence, is this even allowed? A function call expression does not have a `selector`, does it?

## Which Contracts are Affected?

> If you still using the legacy code generation pipeline and your contract uses the 
> ``.selector`` member access on some complex expressions,
> like the return of a function call, such expressions will not be evaluated in the 
> legacy code generation, leading to a different behavior when compiling with and without ``via-ir``.

Reviewer: ```suggestion
If you are still using the legacy code generation pipeline and your contract uses the 
``.selector`` member access on some complex expressions,
like the return value of a function call, such expressions will not be evaluated in the 
legacy code generation, leading to a different behavior when compiling with and without ``via-ir``.
```

Also, this mostly repeats what was already said in the intro. I think you should instead list the specific conditions needed to trigger it.

For example, the code below, if compiled with ``via-ir`` will result in ``x`` being 

> assigned ``42``, evaluating the function call ``h()`` expression.

Reviewer: ```suggestion
assigned ``42``, as a side-effect of evaluating the function call ``h()``.
```

> On the other hand, if compiled without ``via-ir`` it will result in ``x`` being ``0``.

Reviewer: Also, the contract `D` won't even get deployed, will it? Maybe it would even be better to make it the main point of the example.

Author: Yes, it won't. I will update the text.
